accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+with our existing baseline compiler Winch.
+
+### Breakpoints and Single-stepping
+
+To implement breakpoints and single-stepping, we need the ability to pause
+program execution and wait for user input prior to executing a specific line of
+code. The way this is handled in native debuggers is as follows:
+
+* When a user sets a breakpoint, the instruction corresponding to that line of
+  code is replaced with an `int3` instruction.
+
+* When executed, the `int3` instruction raises a `SIGTRAP`, halting execution
+  and yielding control back to the debugger.
+
+This approach leads to a problem for us: we only have one compiled copy of a
+module’s code, even though multiple instances may be running the code. So if we
+just replace an instruction in the code with an interrupt, it is possible that a
+different instance will hit that interrupt. At that point, in order for that
+instance to continue executing we need to swap in the actual instruction. This
+leads to a race condition between the instances where we aren’t guaranteed that
+our breakpoint will actually interrupt when we need it to! We can get around
+this in the following way:
+
+1. When code is compiled for debugging, for every instruction emitted, emit a
+call to a utility function that can check if execution should be paused:
+
+2. When a breakpoint is set, the function will pause execution, allowing for
+resumption later on
+
+3. When an instance hits that breakpoint it can automatically continue if it
+isn’t intended to stop, otherwise it can wait for user input.
+
+We will need to keep track of some information for every wasm opcode in our
+program: what program counter the opcode is, the size of its operand stack, and
+which entry in the operand stack the opcode is in terms of machine state. All of
+this information can be serialized and passed into the utility function present
+in the `<check for breakpoint>` pseudo opcodes above.
+
+### Watchpoints
+
+Watchpoints for locals, globals, and table elements can be converted to
+breakpoints by statically checking for local.set, global.set, and table.set
+instructions. Memory watchpoints— pausing execution when a program writes to a
+specific memory location— are a little bit trickier to accomplish. However, like
+the proposed implementation for single-stepping mentioned in the [previous
+section](#breakpoints-and-single-stepping), we can implement watch-points by
+instrumenting all store operations with a similar call out to a utility
+function. The call sites (store instructions) would have enough local
+information available to determine if a watchpoint would trigger, and could
+provide us with another implementation path that would avoid modifying the
+signal handler.
+
+### Step into / out of
+
+Stepping into a function means pausing execution on the first instruction
+_inside_ a function that is about to be called. We can accomplish this by
+setting a temporary breakpoint on that first instruction, but first we need to
```

Comment:
You can also use the processor's single stepping support to step until right after the call instruction.

Reply:
That's true, but it does require us to have a bunch of per-architecture logic. I suppose the same could be said for the instrumentation-based approach, in that the instrumentation will be different snippets of machine instructions for each different architecture, but at least there the semantics of each snippet of instrumentation as a whole are identical and we don't need to worry about subtle differences that may exist between different architecture's single-stepping support.

Ultimately, I personally expect that the instrumentation approach will be easier to implement correctly (I'm not going to stop harping on this point :-p) than relying on hardware single stepping.

Reply:
I though PTRACE_SINGLESTEP was universally supported, but it seems that 32bit arm doesn't support it architecturally and the emulation support was removed in https://lore.kernel.org/linux-arm-kernel/1297248175-11952-1-git-send-email-will.deacon@arm.com/